src/uct/cuda/gdr_copy/gdr_copy_md.c

```diff
@@ -259,36 +300,246 @@ static uct_md_ops_t md_ops = {
+                                                          uct_gdr_copy_rcache_region_t);
+    uct_gdr_copy_mem_t *memh = &region->memh;
+
+    snprintf(buf, max, "bar ptr:%p", memh->bar_ptr);
+}
+
+static ucs_rcache_ops_t uct_gdr_copy_rcache_ops = {
+    .mem_reg     = uct_gdr_copy_rcache_mem_reg_cb,
+    .mem_dereg   = uct_gdr_copy_rcache_mem_dereg_cb,
+    .merge       = (void*)ucs_empty_function,
+    .dump_region = uct_gdr_copy_rcache_dump_region_cb
+};
+
+static ucs_status_t
+uct_gdr_copy_md_init(uct_gdr_copy_md_t *md, uct_component_t *component,
+                     const uct_gdr_copy_md_config_t *md_config, int locked)
+{
+    ucs_rcache_params_t rcache_params;
+    ucs_status_t status;
+
     md->super.component = &uct_gdr_copy_component;
     md->reg_cost        = md_config->uc_reg_cost;
 
+    if (locked) {
+        md->super.ops = &md_locked_ops;
+    } else {
+        md->super.ops = &md_ops;
+    }
+
     md->gdrcpy_ctx = gdr_open();
     if (md->gdrcpy_ctx == NULL) {
         ucs_error("failed to open gdr copy");
-        status = UCS_ERR_IO_ERROR;
-        goto err_free_md;
+        return UCS_ERR_IO_ERROR;
+    }
+
+    if (md_config->enable_rcache != UCS_NO) {
+        ucs_rcache_set_default_params(&rcache_params);
+        rcache_params.region_struct_size = sizeof(uct_gdr_copy_rcache_region_t);
+        rcache_params.ucm_events         = UCM_EVENT_MEM_TYPE_FREE;
+        rcache_params.context            = md;
+        rcache_params.ops                = &uct_gdr_copy_rcache_ops;
+        rcache_params.flags              = 0;
+
+        status = ucs_rcache_create(&rcache_params, "gdr_copy", NULL,
+                                   &md->rcache);
+        if (status == UCS_OK) {
+            md->super.ops = &md_rcache_ops;
+            md->reg_cost  = UCS_LINEAR_FUNC_ZERO;
+        } else {
+            if (md_config->enable_rcache == UCS_YES) {
+                status = UCS_ERR_IO_ERROR;
+                goto err_close_gdr;
+            } else {
+                ucs_debug("could not create registration cache for: %s",
+                          ucs_status_string(status));
+            }
+        }
     }
 
-    *md_p = (uct_md_h) md;
     return UCS_OK;
 
-err_free_md:
-    ucs_free(md);
+err_close_gdr:
+    gdr_close(md->gdrcpy_ctx);
+    return status;
+}
+
+static ucs_status_t
+uct_gdr_copy_md_open(uct_component_t *component, const char *md_name,
+                     const uct_md_config_t *config, uct_md_h *md_p)
+{
+    const uct_gdr_copy_md_config_t *md_config =
+                    ucs_derived_of(config, uct_gdr_copy_md_config_t);
+    ucs_status_t status;
+    uct_gdr_copy_md_t *md;
+
+    if (!md_config->single) {
+        md = ucs_malloc(sizeof(uct_gdr_copy_md_t), "uct_gdr_copy_md_t");
+        if (NULL == md) {
+            ucs_error("failed to allocate memory for uct_gdr_copy_md_t");
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        status = uct_gdr_copy_md_init(md, component, md_config, 0);
+
```

Comment:
1. remove space line
2. maybe make uct_gdr_copy_context.md a pointer (to not consume memory when gdrcopy not used) and move ucs_malloc inside uct_gdr_copy_md_init (rename it to uct_gdr_copy_md_create or uct_gdr_copy_md_new)